service/history/queues/queue_base.go

```diff
@@ -372,54 +371,71 @@ func (p *queueBase) checkpoint() {
 			continue
 		}
 
-		totalSlices += len(scopes)
 		readerScopes[readerID] = scopes
 		for _, scope := range scopes {
-			newInclusiveLowWatermark = tasks.MinKey(newInclusiveLowWatermark, scope.Range.InclusiveMin)
+			newExclusiveDeletionHighWatermark = tasks.MinKey(newExclusiveDeletionHighWatermark, scope.Range.InclusiveMin)
 		}
 	}
 	p.metricsHandler.Histogram(QueueReaderCountHistogram, metrics.Dimensionless).Record(int64(len(readerScopes)))
-	p.metricsHandler.Histogram(QueueSliceCountHistogram, metrics.Dimensionless).Record(int64(totalSlices))
+	p.metricsHandler.Histogram(QueueSliceCountHistogram, metrics.Dimensionless).Record(int64(p.monitor.GetTotalSliceCount()))
+	p.metricsHandler.Histogram(PendingTasksCounter, metrics.Dimensionless).Record(int64(p.monitor.GetTotalPendingTaskCount()))
 
 	// NOTE: Must range complete task first.
 	// Otherwise, if state is updated first, later deletion fails and shard get reloaded
 	// some tasks will never be deleted.
-	if newInclusiveLowWatermark != tasks.MaximumKey && newInclusiveLowWatermark.CompareTo(p.inclusiveLowWatermark) > 0 {
-		p.metricsHandler.Counter(TaskBatchCompleteCounter).Record(1)
-
-		persistenceMinTaskKey := p.inclusiveLowWatermark
-		persistenceMaxTaskKey := newInclusiveLowWatermark
-		if p.category.Type() == tasks.CategoryTypeScheduled {
-			persistenceMinTaskKey = tasks.NewKey(p.inclusiveLowWatermark.FireTime, 0)
-			persistenceMaxTaskKey = tasks.NewKey(newInclusiveLowWatermark.FireTime, 0)
+	//
+	// Emit metric before the deletion watermark comparsion so we have the emit even if there's no task
+	// for the queue
+	p.metricsHandler.Counter(TaskBatchCompleteCounter).Record(1)
+	if newExclusiveDeletionHighWatermark.CompareTo(p.exclusiveDeletionHighWatermark) > 0 {
+		err = p.rangeCompleteTasks(p.exclusiveDeletionHighWatermark, newExclusiveDeletionHighWatermark)
+		if err != nil {
+			return
 		}
 
-		ctx, cancel := newQueueIOContext()
-		defer cancel()
+		p.exclusiveDeletionHighWatermark = newExclusiveDeletionHighWatermark
+	}
 
-		if err = p.shard.GetExecutionManager().RangeCompleteHistoryTasks(ctx, &persistence.RangeCompleteHistoryTasksRequest{
-			ShardID:             p.shard.GetShardID(),
-			TaskCategory:        p.category,
-			InclusiveMinTaskKey: persistenceMinTaskKey,
-			ExclusiveMaxTaskKey: persistenceMaxTaskKey,
-		}); err != nil {
-			p.logger.Error("Error range completing queue task", tag.Error(err))
-			return
-		}
+	err = p.updateQueueState(readerScopes)
```

Comment:
I would use named return value (retErr), and do return p.updateQueueState() here. Assign to err at last line of func is weird, but it is used in defer func.